.github/workflows/build-test-ship.yaml

```diff
@@ -183,34 +190,251 @@ jobs:
+      # from ECR (https://app.asana.com/0/1199932090708798/1202699911962363/f),
+      # this step will be unnecessary.
+      - name: Use latest service images from ECR
+        run: sed -i 's/${USER:?USER}/${{ steps.login-ecr.outputs.registry }}/' docker-compose.yml
+
+      - name: Create docker-compose override and AWS credentials files
+        env:
+          ENCODED_CREDENTIALS: ${{ secrets.DEV_AWS_CREDENTIALS }}
+        shell: bash
+        run: |
+          yq -n ".version = \"3.7\"" > docker-compose.override.yaml
+          for service in ${{ steps.get-services.outputs.services }}; do
+            yq -i ".services.${service}.image = \"${service}:ci\"" docker-compose.override.yaml
+          done
+          yq eval-all --inplace 'select(fileIndex == 0) * select(fileIndex == 1)' docker-compose.override.yaml ${{ github.workspace }}/docker-compose.ci-override.yml
+          echo docker-compose.override.yaml:
+          cat docker-compose.override.yaml
+          echo $ENCODED_CREDENTIALS | base64 -di > vault/aws_credentials
+
+      # TODO: This is a hack to place empty secret files in the vault directory
+      # (to satisfy docker-compose.yml requirements) without actually retrieving
+      # them from vault.
+      - name: Retrieve vault secrets
+        shell: bash
+        run: |
+          touch vault/website.secrets.env
+          touch vault/landingpageadmin.secrets.env
+
+      - name: Parse service dependencies
+        id: parse-deps
+        run: |
+          # Traverse depends_on values to build the unified dependency set.
+          declare -A work_set # Nodes to be visited
+          declare -A deps # Nodes already in the dependency set
+          # Initialize the work set to the services from this repo.
+          for service in ${{ steps.get-services.outputs.services }}; do
+            work_set[${service}]=1
+          done
+          # While the work set is nonempty, pop and visit an arbitrary node.
+          while [[ ${#work_set[@]} -gt 0 ]]; do
+            keys=(${!work_set[@]})
+            next=${keys[0]}
+            unset work_set[${next}]
+            # Add the next node to the dependency set, and add any of its
+            # unvisited dependencies to the work set.
+            deps[${next}]=1
+            next_deps=$(yq ".services.${next}.depends_on[]" docker-compose.yml)
+            for next_dep in ${next_deps}; do
+              if [[ ! ${deps[${next_dep}]} ]]; then
+                work_set[${next_dep}]=1
+              fi
+            done
+          done
+          echo Dependencies: ${!deps[@]}
+          echo "deps=${!deps[@]}" >> $GITHUB_OUTPUT
+
+      - name: Start postgres and create service databases and roles
+        shell: bash
+        run: |
+          docker-compose up -d postgres
+          timeout 1m bash -c 'until docker-compose run postgres pg_isready -d postgres://${POSTGRES_USER}:${POSTGRES_PASSWORD}@postgres; do sleep 1s; done'
```

Comment:
This should be exactly the kind of thing that [healthchecks](https://docs.docker.com/compose/compose-file/compose-file-v3/#healthcheck) and the [`--wait` option](https://docs.docker.com/engine/reference/commandline/compose_up/#options) of `docker compose up` are for. However, when I tried that ([`actions` PR](https://github.com/nicheinc/actions/pull/8/files), [`development` branch](https://github.com/nicheinc/development/compare/master...x/pg-healthcheck)), `docker compose` appeared to [give up immediately](https://nicheinc.slack.com/archives/C026TLM8HSL/p1666027836893279?thread_ts=1666016597.341929&cid=C026TLM8HSL) upon seeing that `postgres` was unhealthy, rather than retrying until it was healthy. (Worse, it failed inconsistently.) Let me know if I was just just holding it wrong.